### PR TITLE
provider: add a AWS Elemental MediaConvert provider

### LINKED ISSUE #194: providers: add support for mediaconvert
AWS seems to be discontinuing Elastic Transcoder in favor of [MediaConvert](https://aws.amazon.com/mediaconvert/). It would be great to see MediaConvert support on this API.

### PR DESCRIPTION
**Pull request by zsiec**
https://docs.aws.amazon.com/mediaconvert/latest/apireference/resources.html

Closes #194

### CODE REVIEW

File: provider/mediaconvert/mediaconvert.go [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+		close(presetCh)
+	}()
+
+	for res := range presetCh {
+		if res.err != nil {
+			return nil, fmt.Errorf("error getting preset info: %s", res.err)
+		}
+
+		presets[res.presetID] = res.preset
+	}
```
Comment (fsouza) [inline provider/mediaconvert/mediaconvert.go]: Could we abstract all of that into a function that returns the presets map? Something like:

```go
func (p *mcProvider) getPresets(job *db.Job) (map[string]mediaconvert.Preset, error) {
    // impl...
}
```

File: provider/mediaconvert/mediaconvert.go [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+				FileGroupSettings: &mediaconvert.FileGroupSettings{
+					Destination: aws.String(destination),
+				},
+			}
+		default:
+			return nil, fmt.Errorf("container %s is not yet supported with mediaconvert", string(container))
+		}
+
+		mcOutputGroups = append(mcOutputGroups, mcOutputGroup)
+	}
```
Comment (fsouza) [inline provider/mediaconvert/mediaconvert.go]: Same here, can probably create a separate function to build the output groups.

File: provider/mediaconvert/mediaconvert_test.go [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+			}
+		})
+	}
+}
+
+func Test_mcProvider_GetPreset(t *testing.T) {
+	presetID := "some_preset"
+	client := &testMediaConvertClient{t: t}
+	p := &mcProvider{client: client}
+	_, _ = p.GetPreset(presetID)
```
Comment (fsouza) [inline provider/mediaconvert/mediaconvert_test.go]:
```suggestion
	p.GetPreset(presetID)
```

File: provider/mediaconvert/mediaconvert_test.go [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+			client := &testMediaConvertClient{t: t}
+			p := &mcProvider{client: client}
+			_, err := p.CreatePreset(tt.presetModifier(defaultPreset))
+			if err != nil && tt.wantErrMsg != err.Error() {
+				t.Errorf("mcProvider.CreatePreset() error = %v, wantErr %q", err, tt.wantErrMsg)
+				return
+			}
+
+			if tt.assertion != nil {
+				tt.assertion(client.createPresetCalledWith, t)
```
Comment (fsouza) [inline provider/mediaconvert/mediaconvert_test.go]: This is an interesting pattern!

File: provider/mediaconvert/mediaconvert_test.go [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+	if g, e := client.getPresetCalledWith, presetID; g != e {
+		t.Fatalf("got %q, expected %q", g, e)
+	}
+}
+
+func Test_mcProvider_DeletePreset(t *testing.T) {
+	presetID := "some_preset_id"
+	client := &testMediaConvertClient{t: t}
+	p := &mcProvider{client: client}
+	_ = p.DeletePreset(presetID)
```
Comment (fsouza) [inline provider/mediaconvert/mediaconvert_test.go]: Can we add an error check here?

File: provider/mediaconvert/mediaconvert_test.go [hunk 6]
```diff
@@ -0,0 +1,643 @@
...
+			}
+		})
+	}
+}
+
+func Test_mcProvider_CancelJob(t *testing.T) {
+	jobID := "some_job_id"
+	client := &testMediaConvertClient{t: t}
+	p := &mcProvider{client: client}
+	_ = p.CancelJob(jobID)
```
Comment (fsouza) [inline provider/mediaconvert/mediaconvert_test.go]: Same here, can we check the error?

File: provider/mediaconvert/mediaconvert_test.go [hunk 7]
```diff
@@ -0,0 +1,643 @@
...
+	if g, e := client.cancelJobCalledWith, jobID; g != e {
+		t.Fatalf("got %q, expected %q", g, e)
+	}
+}
+
+func Test_mcProvider_Healthcheck(t *testing.T) {
+	client := &testMediaConvertClient{t: t}
+	p := &mcProvider{client: client}
+
+	_ = p.Healthcheck()
```
Comment (fsouza) [inline provider/mediaconvert/mediaconvert_test.go]: Error check here too

File: provider/mediaconvert/mediaconvert_test.go [hunk 8]
```diff
@@ -0,0 +1,643 @@
...
+
+			if g, e := status, &tt.wantStatus; !reflect.DeepEqual(g, e) {
+				t.Fatalf("mcProvider.JobStatus(): wrong job request\nWant %+v\nGot %+v\nDiff %s", e,
+					g, cmp.Diff(e, g))
+			}
+		})
+	}
+}
+
+type testMediaConvertClient struct {
```
Comment (fsouza) [inline provider/mediaconvert/mediaconvert_test.go]: Can you move the test client to a separate file?

File: provider/mediaconvert/preset_mapping.go [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+	mode = strings.ToLower(mode)
+	switch mode {
+	case "", "progressive":
+		return mediaconvert.H264InterlaceModeProgressive, nil
+	default:
+		return "", fmt.Errorf("h264 interlace mode %q is not supported with mediaconvert", mode)
+	}
+}
+
+func videoPresetFrom(preset db.Preset) (*mediaconvert.VideoDescription, error) {
```
Comment (fsouza) [inline provider/mediaconvert/preset_mapping.go]: Could we add some tests for this and ``audioPresetFrom``?

File: README.md [hunk 10]
```diff
@@ -94,6 +95,18 @@ export ZENCODER_API_KEY=your.api.key
 export ZENCODER_DESTINATION=http://access.key.id:secret.access.key@your-s3-bucket.s3.amazonaws.com/
 ```
 
+#### For [MediaConvert](https://aws.amazon.com/mediaconvert/)
+
+```
+export MEDIACONVERT_AWS_ACCESS_KEY_ID=your.access.key.id
+export MEDIACONVERT_AWS_SECRET_ACCESS_KEY=your.secret.access.key
```
[the remark this answers is not in the document]
Comment (zsiec) [reply README.md]: Good question, I thought about this a little and wondered if that would be unexpected to users given the existing patterns, but agree it'd be a nice feature. It'd be trivial to support this here. Let me know if you'd like to support this with MediaConvert.

File: provider/mediaconvert/mediaconvert.go [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+
+const (
+	// Name identifies the MediaConvert provider by name
+	Name = "mediaconvert"
+
+	defaultAudioSampleRate = 48000
+)
+
+func init() {
+	_ = provider.Register(Name, mediaconvertFactory)
```
[the remark this answers is not in the document]
Comment (zsiec) [reply provider/mediaconvert/mediaconvert.go]: 👍

File: provider/mediaconvert/mediaconvert.go (hunk 1 above)
Comment (zsiec) [reply provider/mediaconvert/mediaconvert.go]: Absolutely, I'll do this.

File: provider/mediaconvert/mediaconvert.go (hunk 2 above)
Comment (zsiec) [reply provider/mediaconvert/mediaconvert.go]: Will do.

File: provider/mediaconvert/mediaconvert_test.go (hunk 8 above)
Comment (zsiec) [reply provider/mediaconvert/mediaconvert_test.go]: Sure thing

File: provider/mediaconvert/preset_mapping.go (hunk 9 above)
Comment (zsiec) [reply provider/mediaconvert/preset_mapping.go]: Do you thing separate tests are really needed for these non-exported methods. This method and `audioPresetFrom ` already have 100% coverage. https://codecov.io/gh/nytimes/video-transcoding-api/pull/223/diff?src=pr&el=tree#D4-123

File: provider/mediaconvert/mediaconvert.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (zsiec) [reply provider/mediaconvert/mediaconvert.go]: done

File: provider/mediaconvert/mediaconvert.go (hunk 1 above)
Comment (zsiec) [reply provider/mediaconvert/mediaconvert.go]: done

File: provider/mediaconvert/mediaconvert_test.go (hunk 3 above)
Comment (zsiec) [reply provider/mediaconvert/mediaconvert_test.go]: done

File: provider/mediaconvert/mediaconvert_test.go (hunk 4 above)
Comment (zsiec) [reply provider/mediaconvert/mediaconvert_test.go]: I was looking for a handy way to test atomic aspects of the mapping, this seemed to work :)

File: provider/mediaconvert/mediaconvert_test.go (hunk 5 above)
Comment (zsiec) [reply provider/mediaconvert/mediaconvert_test.go]: done

File: provider/mediaconvert/mediaconvert_test.go (hunk 6 above)
Comment (zsiec) [reply provider/mediaconvert/mediaconvert_test.go]: done

File: provider/mediaconvert/mediaconvert_test.go (hunk 7 above)
Comment (zsiec) [reply provider/mediaconvert/mediaconvert_test.go]: done

File: provider/mediaconvert/mediaconvert_test.go (hunk 8 above)
Comment (zsiec) [reply provider/mediaconvert/mediaconvert_test.go]: done

File: provider/mediaconvert/preset_mapping.go (hunk 9 above)
Comment (fsouza) [reply provider/mediaconvert/preset_mapping.go]: That's a good point, let's leave it as is!

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (fsouza) [reply README.md]: Yeah let's do that, then I think it's good to merge! Or let me know if you'd rather have it merged and work on it on a follow-up PR.

[the remark this answers is not in the document]
Comment (zsiec) [reply README.md]: 👍Will do this today as part of this PR, after I get back home.

File: provider/mediaconvert/testing.go [hunk 12]
```diff
@@ -0,0 +1,102 @@
+package mediaconvert
```
Comment (fsouza) [inline provider/mediaconvert/testing.go]: Just need to name this file with a ``_test`` suffix (something like ``fake_client_test.go``, maybe?). Just to make sure it doesn't get compiled with the production binaries (I know the linker will likely eliminate it as dead code, but it's good to be safe!)

Comment (zsiec) [reply provider/mediaconvert/testing.go]: Will do

Comment (zsiec) [reply provider/mediaconvert/testing.go]: done

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (zsiec) [reply README.md]: Just updated this.

**Pull request merged**